File: LLD/week2/patterns/singleton/basic_example.py
"""
Basic Singleton Implementation

This module demonstrates a simple singleton pattern using Python.
Shows both eager and lazy initialization approaches.
"""


class DatabaseConnection:
    """
    Simple singleton for database connection management.

    Uses lazy initialization - instance created only when first requested.
    Not thread-safe - suitable for single-threaded applications only.
    """

    _instance = None

    def __init__(self):
        """Private constructor - should not be called directly."""
        if DatabaseConnection._instance is not None:
            raise RuntimeError(
                "Singleton instance already exists! Use get_instance() instead."
            )

        # Initialize connection details
        self.host = "localhost"
        self.port = 5432
        self.database = "myapp_db"
        self.connected = False
        self.connection_count = 0

    @classmethod
    def get_instance(cls):
        """
        Get the singleton instance.

        Returns:
            DatabaseConnection: The singleton instance
        """
        if cls._instance is None:
            cls._instance = DatabaseConnection()
            print(f"Created new DatabaseConnection instance: {id(cls._instance)}")
        return cls._instance

    def connect(self):
        """Simulate connecting to database."""
        if not self.connected:
            self.connected = True
            self.connection_count += 1
            print(f"Connected to {self.host}:{self.port}/{self.database}")
        else:
            print("Already connected")

    def disconnect(self):
        """Simulate disconnecting from database."""
        if self.connected:
            self.connected = False
            print(f"Disconnected from database")

    def execute_query(self, query: str):
        """Execute a database query."""
        if not self.connected:
            raise RuntimeError("Not connected to database")
        print(f"Executing query: {query}")
        return f"Result of: {query}"


class ConfigurationManager:
    """
    Singleton configuration manager with eager initialization.

    Instance created immediately when class loads.
    Thread-safe by default (Python module loading is thread-safe).
    """

    # Eager initialization - instance created immediately
    _instance = None

    def __init__(self):
        """Private constructor."""
        if ConfigurationManager._instance is not None:
            raise RuntimeError("Use get_instance() to access ConfigurationManager")

        # Load configuration
        self.config = {
            "app_name": "MyApplication",
            "version": "1.0.0",
            "debug": True,
            "max_connections": 100,
            "timeout": 30
        }
        print("Configuration loaded")

    @classmethod
    def get_instance(cls):
        """Get the singleton instance."""
        if cls._instance is None:
            cls._instance = ConfigurationManager()
        return cls._instance

    def get(self, key: str, default=None):
        """Get configuration value."""
        return self.config.get(key, default)

    def set(self, key: str, value):
        """Set configuration value."""
        self.config[key] = value
        print(f"Config updated: {key} = {value}")


def demonstrate_basic_singleton():
    """Demonstrate basic singleton behavior."""
    print("="*70)
    print("BASIC SINGLETON DEMONSTRATION")
    print("="*70)

    print("\n1. Getting first database connection instance:")
    db1 = DatabaseConnection.get_instance()
    print(f"   Instance ID: {id(db1)}")

    print("\n2. Getting second database connection instance:")
    db2 = DatabaseConnection.get_instance()
    print(f"   Instance ID: {id(db2)}")

    print(f"\n3. Checking if instances are identical:")
    print(f"   db1 is db2: {db1 is db2}")
    print(f"   Same memory address: {id(db1) == id(db2)}")

    print("\n4. Connecting via first instance:")
    db1.connect()

    print("\n5. Checking connection status via second instance:")
    print(f"   db2.connected: {db2.connected}")
    print("   (State is shared because it's the same instance!)")

    print("\n6. Attempting to create instance directly:")
    try:
        db3 = DatabaseConnection()
    except RuntimeError as e:
        print(f"   Error (expected): {e}")


def demonstrate_configuration_singleton():
    """Demonstrate configuration manager singleton."""
    print("\n" + "="*70)
    print("CONFIGURATION SINGLETON DEMONSTRATION")
    print("="*70)

    print("\n1. Getting first config instance:")
    config1 = ConfigurationManager.get_instance()
    print(f"   App Name: {config1.get('app_name')}")
    print(f"   Instance ID: {id(config1)}")

    print("\n2. Modifying config via first instance:")
    config1.set("debug", False)

    print("\n3. Getting second config instance:")
    config2 = ConfigurationManager.get_instance()
    print(f"   Instance ID: {id(config2)}")
    print(f"   Debug mode: {config2.get('debug')}")
    print("   (Change visible through second reference!)")

    print("\n4. Verifying instances are identical:")
    print(f"   config1 is config2: {config1 is config2}")


if __name__ == "__main__":
    demonstrate_basic_singleton()
    demonstrate_configuration_singleton()

    print("\n" + "="*70)
    print("KEY POINTS")
    print("="*70)
    print("""
1. Only one instance exists for the entire application
2. All references point to the same object in memory
3. State changes are visible through all references
4. Direct instantiation is prevented (raises error)
5. Access controlled through get_instance() class method
    """)

File: LLD/week2/patterns/singleton/basic_example.java
/**
 * Basic Singleton Implementation
 *
 * This file demonstrates a simple singleton pattern using Java.
 * Shows both eager and lazy initialization approaches.
 */

import java.util.HashMap;
import java.util.Map;

/**
 * Simple singleton for database connection management.
 *
 * Uses lazy initialization - instance created only when first requested.
 * Not thread-safe - suitable for single-threaded applications only.
 */
class DatabaseConnection {
    private static DatabaseConnection instance = null;

    // Connection details
    private String host;
    private int port;
    private String database;
    private boolean connected;
    private int connectionCount;

    /**
     * Private constructor - prevents external instantiation.
     */
    private DatabaseConnection() {
        this.host = "localhost";
        this.port = 5432;
        this.database = "myapp_db";
        this.connected = false;
        this.connectionCount = 0;
    }

    /**
     * Get the singleton instance.
     *
     * @return The singleton instance
     */
    public static DatabaseConnection getInstance() {
        if (instance == null) {
            instance = new DatabaseConnection();
            System.out.println("Created new DatabaseConnection instance: " +
                             System.identityHashCode(instance));
        }
        return instance;
    }

    /**
     * Simulate connecting to database.
     */
    public void connect() {
        if (!connected) {
            connected = true;
            connectionCount++;
            System.out.println("Connected to " + host + ":" + port + "/" + database);
        } else {
            System.out.println("Already connected");
        }
    }

    /**
     * Simulate disconnecting from database.
     */
    public void disconnect() {
        if (connected) {
            connected = false;
            System.out.println("Disconnected from database");
        }
    }

    /**
     * Execute a database query.
     */
    public String executeQuery(String query) {
        if (!connected) {
            throw new RuntimeException("Not connected to database");
        }
        System.out.println("Executing query: " + query);
        return "Result of: " + query;
    }

    public boolean isConnected() {
        return connected;
    }

    public int getConnectionCount() {
        return connectionCount;
    }
}


/**
 * Singleton configuration manager with eager initialization.
 *
 * Instance created immediately when class loads.
 * Thread-safe by default (Java class loading is thread-safe).
 */
class ConfigurationManager {
    // Eager initialization - instance created immediately
    private static ConfigurationManager instance = null;

    private Map<String, String> config;

    /**
     * Private constructor.
     */
    private ConfigurationManager() {
        // Load configuration
        config = new HashMap<>();
        config.put("app_name", "MyApplication");
        config.put("version", "1.0.0");
        config.put("debug", "true");
        config.put("max_connections", "100");
        config.put("timeout", "30");
        System.out.println("Configuration loaded");
    }

    /**
     * Get the singleton instance.
     */
    public static ConfigurationManager getInstance() {
        if (instance == null) {
            instance = new ConfigurationManager();
        }
        return instance;
    }

    /**
     * Get configuration value.
     */
    public String get(String key) {
        return config.getOrDefault(key, null);
    }

    /**
     * Get configuration value with default.
     */
    public String get(String key, String defaultValue) {
        return config.getOrDefault(key, defaultValue);
    }

    /**
     * Set configuration value.
     */
    public void set(String key, String value) {
        config.put(key, value);
        System.out.println("Config updated: " + key + " = " + value);
    }
}


/**
 * Demonstration class for basic singleton patterns.
 */
public class basic_example {

    public static void demonstrateBasicSingleton() {
        System.out.println("=".repeat(70));
        System.out.println("BASIC SINGLETON DEMONSTRATION");
        System.out.println("=".repeat(70));

        System.out.println("\n1. Getting first database connection instance:");
        DatabaseConnection db1 = DatabaseConnection.getInstance();
        System.out.println("   Instance ID: " + System.identityHashCode(db1));

        System.out.println("\n2. Getting second database connection instance:");
        DatabaseConnection db2 = DatabaseConnection.getInstance();
        System.out.println("   Instance ID: " + System.identityHashCode(db2));

        System.out.println("\n3. Checking if instances are identical:");
        System.out.println("   db1 == db2: " + (db1 == db2));
        System.out.println("   Same memory address: " +
                         (System.identityHashCode(db1) == System.identityHashCode(db2)));

        System.out.println("\n4. Connecting via first instance:");
        db1.connect();

        System.out.println("\n5. Checking connection status via second instance:");
        System.out.println("   db2.isConnected(): " + db2.isConnected());
        System.out.println("   (State is shared because it's the same instance!)");

        System.out.println("\n6. Cannot create instance directly (prevented by private constructor)");
        System.out.println("   Uncommenting 'DatabaseConnection db3 = new DatabaseConnection();' would cause compilation error");
        // DatabaseConnection db3 = new DatabaseConnection();  // Compilation error
    }

    public static void demonstrateConfigurationSingleton() {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("CONFIGURATION SINGLETON DEMONSTRATION");
        System.out.println("=".repeat(70));

        System.out.println("\n1. Getting first config instance:");
        ConfigurationManager config1 = ConfigurationManager.getInstance();
        System.out.println("   App Name: " + config1.get("app_name"));
        System.out.println("   Instance ID: " + System.identityHashCode(config1));

        System.out.println("\n2. Modifying config via first instance:");
        config1.set("debug", "false");

        System.out.println("\n3. Getting second config instance:");
        ConfigurationManager config2 = ConfigurationManager.getInstance();
        System.out.println("   Instance ID: " + System.identityHashCode(config2));
        System.out.println("   Debug mode: " + config2.get("debug"));
        System.out.println("   (Change visible through second reference!)");

        System.out.println("\n4. Verifying instances are identical:");
        System.out.println("   config1 == config2: " + (config1 == config2));
    }

    public static void main(String[] args) {
        demonstrateBasicSingleton();
        demonstrateConfigurationSingleton();

        System.out.println("\n" + "=".repeat(70));
        System.out.println("KEY POINTS");
        System.out.println("=".repeat(70));
        System.out.println("""

1. Only one instance exists for the entire application
2. All references point to the same object in memory
3. State changes are visible through all references
4. Direct instantiation is prevented (raises error)
5. Access controlled through getInstance() static method
        """);
    }
}
